In [55]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

In [56]:
data = pd.read_csv('glcm_result.csv')
data.head(6)

Unnamed: 0                                              path   contrast_0  \
0           0                        datasets/ayam\download.jpg  1952.343750   
1           1  datasets/ayam\egg-white-food-protein-162712.jpeg  2828.906250   
2           2                      datasets/ayam\images (1).jpg  2346.796875   
3           3                          datasets/ayam\images.jpg  1988.203125   
4           4             datasets/ayam\IMG_20220818_154051.jpg  3882.773438   
5           5             datasets/ayam\IMG_20220818_160129.jpg  2020.078125   

   dissimilarity_0  homogeneity_0  energy_0  correlation_0     entropy_0  \
0         7.656250       0.970093  0.866311       0.712865 -5.105249e+06   
1        11.093750       0.956665  0.816831       0.692669 -7.070763e+06   
2         9.203125       0.964050  0.822255       0.747541 -7.147066e+06   
3         7.796875       0.969543  0.866829       0.704665 -5.051554e+06   
4        15.226562       0.940521  0.703014       0.752067 -1.291644e+07   
5         7.921875       0.969055  0.826787       0.782924 -7.155544e+06   

      asm_0  contrast_45  ...    entropy_95    asm_90  contrast_135  \
0  0.000013       3084.0  ... -5.105249e+06  0.000013        3044.0   
1  0.000012       4408.0  ... -7.070763e+06  0.000013        4596.0   
2  0.000013       3870.0  ... -7.147066e+06  0.000012        4188.0   
3  0.000013       3256.0  ... -5.051554e+06  0.000013        3220.0   
4  0.000011       8468.0  ... -1.291644e+07  0.000010        8506.0   
5  0.000013       4668.0  ... -7.155544e+06  0.000012        4556.0   

   dissimilarity_135  homogeneity_135  energy_135  correlation_135  \
0          11.937255         0.953370    0.850427         0.553960   
1          18.023529         0.929596    0.792249         0.502481   
2          16.423529         0.935846    0.796272         0.551083   
3          12.627451         0.950674    0.849020         0.523450   
4          33.356863         0.869700    0.644671         0.458630   
5          17.866667         0.930208    0.791218         0.512167   

    entropy_135   asm_135  labels  
0 -5.105249e+06  0.000013    ayam  
1 -7.070763e+06  0.000012    ayam  
2 -7.147066e+06  0.000012    ayam  
3 -5.051554e+06  0.000013    ayam  
4 -1.291644e+07  0.000010    ayam  
5 -7.155544e+06  0.000012    ayam  

[6 rows x 31 columns]

kode di atas merupakan kode yang digunakan untuk membaca file csv dengan menggunakan liblary pandas. file tersebut dikonversi menjadi objek dataframe.
dan selanjutnya menampilkan 6 baris pertama dari dataframe 'data'.

In [57]:
y = data['labels']
x = data.drop(['path', 'labels', 'Unnamed: 0'], axis=1)

kode di atas merupakan pemilihan 'labels' dari dataframe 'data' yang disimpan ke dalam variabel y. dan selanjutnya data frame baru dengan variabel x dibuat dengan menghapus kolom 'path' dan 'labels' menggunakan fungsi drop dengan parameter kolom yang akan dihapus dan 'axis=1' yang berarti kita menghapus kolom.

In [58]:
data.isnull().any()

Unnamed: 0           False
path                 False
contrast_0           False
dissimilarity_0      False
homogeneity_0        False
energy_0             False
correlation_0        False
entropy_0            False
asm_0                False
contrast_45          False
dissimilarity_45     False
homogeneity_45       False
energy_45            False
correlation_45       False
entropy_45           False
asm_45               False
contrast_90          False
dissimilarity_90     False
homogeneity_90       False
energy_90            False
correlation_90       False
entropy_95           False
asm_90               False
contrast_135         False
dissimilarity_135    False
homogeneity_135      False
energy_135           False
correlation_135      False
entropy_135          False
asm_135              False
labels               False
dtype: bool

kode di atas digunakan untuk melakukan pengecekan terhadap dataframe apakah terdapat nilai kosong pada dataframe. jika sel tersebut bernilai null maka akan diisi dengan nilai true dan False jika sel tersebut memiliki nilai yang valid.
'any()'adalah metode yang digunakan pada DataFrame untuk mengecek apakah ada setidaknya satu nilai True pada setiap kolom. Jika ada nilai True dalam suatu kolom, maka 'any()' akan mengembalikan True untuk kolom tersebut, dan jika semua nilai dalam kolom adalah False, maka 'any()' akan mengembalikan False untuk kolom tersebut.

In [59]:
data = data.fillna(0)
data.isnull().any()

Unnamed: 0           False
path                 False
contrast_0           False
dissimilarity_0      False
homogeneity_0        False
energy_0             False
correlation_0        False
entropy_0            False
asm_0                False
contrast_45          False
dissimilarity_45     False
homogeneity_45       False
energy_45            False
correlation_45       False
entropy_45           False
asm_45               False
contrast_90          False
dissimilarity_90     False
homogeneity_90       False
energy_90            False
correlation_90       False
entropy_95           False
asm_90               False
contrast_135         False
dissimilarity_135    False
homogeneity_135      False
energy_135           False
correlation_135      False
entropy_135          False
asm_135              False
labels               False
dtype: bool

kode di atas digunakan untuk mengelola data yang kosong. fungsi 'data.fillna(0)' digunakan untuk menggati semua data yang kosong dalam dataframe menjadi nilai 0.
selanjutan dilakukan lagi pemerikasaan apakah masih ada data yang kosong pada dataframe.

In [60]:
rf_model = RandomForestClassifier()
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

kode di atas merupakan penginisialisasian objek model klasifikasi random forest dengan menggunakan parameter default. yang berarti berarti model akan dibuat dengan pengaturan default yang telah ditentukan untuk RandomForestClassifier.
selanjutnya merupakan pembagian data menjadi data test ata uji dan train atau pelatihan. data dibagi menjadi 2 bagian berdasarkan variabel 'x' fitur dam 'y' labels. 0.2 atau 20% data akan dialokasikan  untuk pengujian dan sisa datanya 80% akan digunakan untuk pelatihan. Parameter 'random_state=42' digunakan untuk mengatur biji acak (random seed) sehingga pembagian data menjadi set pelatihan dan pengujian dapat direproduksi dengan cara yang sama setiap kali kode tersebut dijalankan.

Setelah membagi data, variabel 'X_train' akan berisi set fitur pelatihan, 'y_train' akan berisi set label pelatihan, 'X_test' akan berisi set fitur pengujian, dan 'y_test' akan berisi set label pengujian. Variabel ini akan digunakan untuk melatih dan menguji model klasifikasi Random Forest yang telah diinisialisasi sebelumnya 'rf_model'.

In [61]:
rf_model.fit(X_train, y_train)
rf_pred= rf_model.predict(X_test)

kode di atas merupakan langkah pelatihan (fitting) model Random Forest. Fungsi 'fit()' digunakan untuk melatih model pada set pelatihan. Variabel 'X_train 'berisi fitur-fitur pelatihan, sedangkan 'y_train' berisi label-label pelatihan. Dengan memanggil .'fit(X_train, y_train)' pada objek model rf_model, model Random Forest akan belajar dari fitur-fitur dan label-label yang telah diberikan.

selanjutnya adalah langkah predikisi menggunakan model random forest yang dilatih. Fungsi 'predict()' digunakan untuk melakukan prediksi pada set pengujian. Variabel 'X_test' berisi fitur-fitur pengujian yang ingin kita lakukan prediksi. Hasil prediksi akan disimpan dalam variabel 'rf_pred'.Hasil prediksi akan disimpan dalam variabel rf_pred untuk digunakan dalam evaluasi kinerja model atau keperluan lainnya.

In [62]:
rf_accuracy = accuracy_score(y_test, rf_pred)
rf_report = classification_report(y_test, rf_pred)

kode di atas digunakan untuk menghitung akurasi model klasifikasi Random Forest dengan membandingkan label aktual 'y_test' dari set pengujian dengan label prediksi 'rf_pred'. Fungsi 'accuracy_score()' dari scikit-learn digunakan untuk menghitung akurasi dengan membandingkan label prediksi dengan label aktual. Nilai akurasi akan disimpan dalam variabel 'rf_accuracy' untuk digunakan atau dicetak nanti.

selanjutnya variabel 'rf_report' berfungsi untuk menghasilkan laporan klasifikasi yang terperinci untuk model klasifikasi Random Forest pada set pengujian. Fungsi 'classification_report()' dari scikit-learn digunakan untuk menghasilkan laporan yang mencakup beberapa metrik evaluasi klasifikasi seperti presisi (precision), recall, nilai F1 (F1-score), dan dukungan (support) untuk setiap kelas. Laporan ini membantu dalam mengevaluasi kinerja model secara komprehensif. Laporan klasifikasi akan disimpan dalam variabel rf_report untuk digunakan atau dicetak nanti.

Dengan menggunakan kode ini, kita dapat mengukur akurasi model klasifikasi Random Forest dengan membandingkan label prediksi dan label aktual menggunakan accuracy_score(). Selain itu, kita juga dapat melihat laporan klasifikasi yang memberikan informasi lebih rinci tentang performa model pada set pengujian menggunakan classification_report().

In [63]:
print("Random Forest Accuracy:", rf_accuracy * 100, "%")
print("Random Forest Classification Report:\n", rf_report)

Random Forest Accuracy: 78.33333333333333 %
Random Forest Classification Report:
               precision    recall  f1-score   support

        ayam       0.83      0.86      0.84        22
      kalkun       0.63      0.75      0.69        16
       puyuh       0.89      0.73      0.80        22

    accuracy                           0.78        60
   macro avg       0.78      0.78      0.78        60
weighted avg       0.80      0.78      0.79        60



kode di atas digunakan untuk mencetak nilai accuracy darim model klasifikasi random forest. nilai akurasi uang di simpan dalam variabel rf_accuracy dikalikan dengan 100 untuk mengubahnya menjadi persentase dan kemudian dicetak. 

dan nilai laporan yang disimpan dalam variabel rf_report akan dicetak dan memberikan laporan  informasi rinci tentang metrik evaluasi klasifikasi seperti presisi, recall, nilai F1, dan dukungan untuk setiap kelas.

# Penjelasan Output
Random Forest Accuracy: 78.33333333333333%: Menunjukkan bahwa akurasi model klasifikasi Random Forest adalah sebesar 78.33%. Akurasi mengukur sejauh mana model dapat melakukan prediksi dengan benar pada set pengujian. Dalam kasus ini, model mencapai tingkat akurasi sebesar 78.33%.

Precision: Menunjukkan presisi (precision) dari model untuk setiap kelas. Presisi adalah ukuran dari sejauh mana prediksi positif yang dilakukan oleh model benar. Semakin tinggi nilai presisi, semakin sedikit prediksi palsu positif yang dibuat oleh model.

Recall: Menunjukkan recall (recall) dari model untuk setiap kelas. Recall adalah ukuran dari sejauh mana model mampu menemukan kembali (mengingat) instans positif yang sebenarnya. Semakin tinggi nilai recall, semakin sedikit instans positif yang terlewatkan oleh model.

F1-Score: Menunjukkan nilai F1-Score untuk setiap kelas. F1-Score adalah ukuran rata-rata harmonik antara presisi dan recall. F1-Score memberikan gambaran keseluruhan tentang keseimbangan antara presisi dan recall.

Support: Menunjukkan jumlah sampel yang ada dalam set pengujian untuk setiap kelas.

Informasi di bawah laporan klasifikasi adalah nilai rata-rata dari presisi, recall, dan F1-Score untuk semua kelas, yaitu "macro avg" dan "weighted avg". "Macro avg" mengambil rata-rata dari metrik evaluasi kelas tanpa memperhatikan ketidakseimbangan jumlah sampel di setiap kelas, sedangkan "weighted avg" memberikan bobot rata-rata berdasarkan jumlah sampel di setiap kelas.